cdap-api-common/src/main/java/io/cdap/cdap/internal/io/SchemaTypeAdapter.java

```diff
@@ -65,74 +66,78 @@ public void write(JsonWriter writer, Schema schema) throws IOException {
       writer.nullValue();
       return;
     }
-    Set<String> knownRecords = new HashSet<>();
-    write(writer, schema, knownRecords);
+    Set<String> definedTypes = new HashSet<>();
```

Comment:
I think we should refactor this class so all of this read and write logic is abstracted and self contained. This way, we don't have to carry this `definedTypes` set across the entire call stack.

Reply:
👍